### PR TITLE
Prompt user for consent upon logging in through google

### PR DESCRIPTION
**Pull request by maddiedierker**
This should fix 2 related Honeybadger errors ([1](https://app.honeybadger.io/projects/3240/faults/34448392#notice-summary), [2](https://app.honeybadger.io/projects/3240/faults/34460333#notice-summary)). I determined that these errors were stemming from the fact that some users do not have a `refresh_token` from Google.

### Problem
We have a lot of Google users who, for whatever reason, are missing a `refresh_token` from Google. This means that once their current `token` expires, the next request to Google fails because there is no `refresh_token` available to grab a new, valid `token`. To get around this, the user has to log out and back in, which updates their `token`, but does not add a `refresh_token` to their account.

### Solution
When a user logs in through Google, explicitly prompt the user for their consent. This will give us a `refresh_token` that we can add to the user's account. 

It isn't _great_ that we will be updating the user's `refresh_token` every time they log in through Google, but I think it's okay for a few reasons:
1. Users stay logged in to Code.org for a long time, so they aren't logging in very often, and
2. Google allows each user-client combination to have 50 valid refresh_tokens, then will begin automatically invalidating old ones: 
> There is currently a limit of 50 refresh tokens per user account per client. If the limit is reached, creating a new refresh token automatically invalidates the oldest refresh token without warning. This limit does not apply to service accounts. ([source](https://developers.google.com/identity/protocols/OAuth2#expiration))

We only care about 1 `refresh_token`, so old tokens being invalidated should not be an issue.

Now, when a user logs in through Google, they will see the consent screen(s) every time:
1. All users:
Screen Shot 2019-03-25 at 9 54 34 AM

2. Users who have granted Google Classroom access:
Screen Shot 2019-03-25 at 9 54 46 AM

This is a minimally invasive inconvenience that will hopefully fix the errors we're seeing (and greatly reduce the number of Zendesk tickets around this issue!)

### CODE REVIEW

File: dashboard/app/controllers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -24,12 +24,14 @@ def clever
 
   # GET /users/auth/google_oauth2/callback
   def google_oauth2
+    user = find_user_by_credential
+    user&.update_oauth_credential_tokens auth_hash
```
Comment (maddiedierker) [inline dashboard/app/controllers/omniauth_callbacks_controller.rb:28]: I snuck this fix in as well: We were only calling `update_oauth_credential_tokens` when users sign in, but we also want to update a user's tokens right after they have given us Google Classroom access. Now, any time a user hits this callback and is logged in, we will update their tokens.

**Comment by Hamms** [comment]
> Users stay logged in to Code.org for a long time, so they aren't logging in very often

Oh, I didn't know that! Where's this information come from?

**Review by Hamms — APPROVED** [review]
nice!

File: dashboard/config/initializers/devise.rb [hunk 2]
```diff
@@ -283,6 +283,7 @@
 
   config.omniauth :google_oauth2, CDO.dashboard_google_key, CDO.dashboard_google_secret, {
     include_granted_scopes: true,
+    prompt: 'consent',
```
Comment (islemaster) [inline dashboard/config/initializers/devise.rb:286]: Does this mean we're going to prompt every user for access to their Google Classroom information, even if they are only using Google for SSO?

**Review by islemaster — APPROVED** [review]

File: dashboard/config/initializers/devise.rb (hunk 2 above)
Comment (maddiedierker) [reply dashboard/config/initializers/devise.rb:286]: nope, only users who have previously given us access to Google Classroom will see the consent page for Google Classroom

**Comment by maddiedierker** [comment]
@Hamms re: users staying logged in for a long time -- i'm not sure there is a great place to point you to, but this is due to the fact that we never log users out unless they explicitly log themselves out

**Comment by Hamms** [comment]
Ah. I understand that _we_ don't log users out, but given that we quite often expect our users to be working on shared devices, shouldn't we in fact expect them to be logging themselves out quite regularly?

**Comment by maddiedierker** [comment]
right, great point. users on shared computers will end up being logged out pretty often. i was thinking about that in relation to Google's "you can only have 50 valid refresh tokens" rule and was concerned at first, but i think it's still okay because Google will 1) invalidate the oldest refresh token (which we don't care about anymore), and 2) still give us a new token. 

it's definitely something for us to watch out for though -- are we seeing issues from users having their refresh tokens updated too often? if they are, i'll have to think of a more clever way to update refresh tokens, but it seems like this won't cause problems at the very least. i asked JoshS to let me know if we're seeing Google wonkiness, so he's aware of the change that's going out

File: dashboard/config/initializers/devise.rb (hunk 2 above)
Comment (joshlory) [reply dashboard/config/initializers/devise.rb:286]: On each login? I'm getting this second prompt every time I log out and log back in:

image

Comment (maddiedierker) [reply dashboard/config/initializers/devise.rb:286]: yes, that will happen on each login for teachers who have previously given consent for Google Classroom access. happy to talk through alternative approaches!

Comment (joshlory) [reply dashboard/config/initializers/devise.rb:286]: Would it be possible to only prompt if the user doesn't have a refresh token already?

Comment (maddiedierker) [reply dashboard/config/initializers/devise.rb:286]: that was my original intent/hope, but there are a few impediments with that:
1. the "consent" step comes before we know anything about the user, so we can't check if we have a refresh token beforehand
2. we can't assume that a refresh token is valid by its presence. refresh tokens are invalidated when a user changes their password or revokes access to code.org then comes back (the latter being very unlikely and not an actual concern)

i think there are a few good alternative options, but they require more time/work:
1. separate the sign in and sign up paths. currently, clicking "continue with google" takes you to a generic route that _decides_ if you're signing in or signing up. if these were separated somehow, and we _knew_ that the user was trying to sign up, we could request consent, whereas if we knew the user was signing in, we would skip the consent step.
2. integrate this consent step into rostered section functionality. when a teacher is signed in and trying to create a google classroom section or sync an existing section, we could automatically detect invalid tokens or a missing refresh token and re-prompt the user for their consent

**Pull request merged**